Validation

svgogin opened the pull request:
1) Validation implemented
2) Questions written

Commit:
#20 Add exception handler
1. Exception handler added
2. Server error messages enabled

Commit:
Merge branch 'main' of https://github.com/svgogin/ServiceRepository into ExceptionHandlers

 Conflicts:
	src/main/resources/application.properties

Commit:
1. Logic of the Post method was transferred from the controller to the service
2. Tests of the exception handler added.
3. Checkstyle issues fixed.
4. Logging of the logic added to the Post method.

Commit:
Add exception handler #20
1. Logic of the Post method was transferred from the controller to the service
2. Tests of the exception handler added.
3. Checkstyle issues fixed.
4. Logging of the logic added to the Post method.

Commit:
Add exception handler #20
1. Logic of the Post method was transferred from the controller to the service
2. Tests of the exception handler added.
3. Checkstyle issues fixed.
4. Logging of the logic added to the Post method.

svgogin committed:
Add exception handler #20
1) Exception handler added
2) Logic moved from the controller to the service
3) Custom exceptions added
4) Tests updated

svgogin committed:
Merge remote-tracking branch 'origin/main' into ExceptionHandlers

# Conflicts:
#	src/main/java/ru/svgogin/service/spark/exception/EntityAlreadyExistsException.java
#	src/main/java/ru/svgogin/service/spark/service/SparkService.java
#	src/main/java/ru/svgogin/service/spark/web/SparkController.java
#	src/test/java/ru/svgogin/service/spark/web/SparkControllerItTest.java

svgogin committed:
Add exception handler #20
1) Code improved according to the review

svgogin committed:
Add exception handler #20
1) Code improved according to the review (code style)

svgogin committed:
Add exception handler #20
1) Imports optimized.

svgogin committed:
Validation #23
1) Validation added
2) Questions written

svgogin committed:
Merge branch 'main' of https://github.com/svgogin/ServiceRepository into Validation

 Conflicts:
	src/main/java/ru/svgogin/service/spark/errordto/ErrorDto.java
	src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java
	src/main/java/ru/svgogin/service/spark/web/SparkController.java

svgogin committed:
Validation #23
3) An issue concerned different response formats solved.

alinkov commented on src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java:
```diff
@@ -35,4 +38,32 @@
+                                                                @NonNull WebRequest request) {
+    var errorField = ex.getBindingResult().getFieldErrors().iterator().next().getField();
+    var message = "Invalid format of request parameter: " + errorField;
+    //toDO If a parameter is not passed in the request, I don't know,
```
please, add tests and also for this case

svgogin replied:
Done

svgogin committed:
Validation #23
4) HandleMethodArgumentNotValid logic improved.
5) Validation tests added
6) Log message formats changed

svgogin committed:
Validation #23
4) HandleMethodArgumentNotValid logic improved.
5) Validation tests added
6) Log message formats changed

alinkov commented on src/main/java/ru/svgogin/service/spark/service/SparkService.java:
```diff
@@ -50,7 +50,7 @@ public CompanyDto update(String inn, CompanyDto companyDto) {
     Company companyForUpdate = actualize(companyDtoEx, companyDto);
     var updatedCompany = aggregateTemplate.update(companyForUpdate);
-    log.info("Company with inn " + inn + " was successfully updated.");
+    log.info("Company with inn {}} was successfully updated.", inn);
```
```suggestion
    log.info("Company with inn {} was successfully updated.", inn);
```

svgogin replied:
Done

alinkov commented on src/main/java/ru/svgogin/service/spark/web/SparkController.java:
```diff
@@ -33,26 +36,36 @@ public SparkController(SparkService sparkService) {
 
   @DeleteMapping("/{inn}")
-  public ResponseEntity<CompanyDto> deleteCompany(@PathVariable("inn") String inn) {
+  public ResponseEntity<CompanyDto> deleteCompany(@PathVariable("inn")
+                                                    @Pattern(regexp = "^(\\d{10}|\\d{12})$",
+                                                             message = "Inn should contain "
+                                                        + "10 or 12 digits") String inn) {
```
It is better to format this way (and in another methods)
```suggestion
  public ResponseEntity<CompanyDto> deleteCompany(
      @PathVariable("inn")
      @Pattern(regexp = "^(\\d{10}|\\d{12})$", message = "Inn should contain 10 or 12 digits") String inn
      ) {
```

svgogin replied:
[INFO] --- maven-checkstyle-plugin:3.1.1:check (validate) @ spark ---
[INFO] Starting audit...
[ERROR] D:\ServiceRepository\src\main\java\ru\svgogin\service\spark\web\SparkController.java:66: Line is longer than 100 characters (found 154). [LineLength]
Audit done.

svgogin replied:
Done

alinkov commented on src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java:
```diff
@@ -35,4 +41,37 @@
+                                                                @NonNull HttpHeaders headers,
+                                                                @NonNull HttpStatus status,
+                                                                @NonNull WebRequest request) {
+    var error = ex.getBindingResult().getFieldErrors().iterator().next();
```
why do you return error only for first field?

alinkov commented on src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java:
```diff
@@ -35,4 +41,37 @@
+                                                                @NonNull WebRequest request) {
+    var error = ex.getBindingResult().getFieldErrors().iterator().next();
+    var errorField = error.getField();
+    //This is a bullshit implementation, but I have no idea, how to improve it
```
it is better to do it this way:
```java
  protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException ex,
                                                                @NonNull HttpHeaders headers,
                                                                @NonNull HttpStatus status,
                                                                @NonNull WebRequest request) {
    var error = ex.getBindingResult().getFieldErrors().iterator().next();
    var errorField = error.getField();
    var errorCode = errorCodeBy(error);
    var message = String.format("%s (%s)", errorCode.message, errorField);
    log.warn(message);
    var bodyOfResponse = new ErrorDto(errorCode, message);
    return handleExceptionInternal(
        ex, bodyOfResponse, headers, status, request
    );
  }

  private ErrorCode errorCodeBy(FieldError fieldError) {
    var notBlankExists = Optional.ofNullable(fieldError.getCodes())
        .map(codes -> Set.of(codes).contains("NotBlank"))
        .orElse(false);
    if (notBlankExists) {
      return ErrorCode.ERROR004;
    } else {
      return ErrorCode.ERROR003;
    }
  }
```
But also you shouldn't return result only for first error

svgogin replied:
Done

alinkov commented on src/test/java/ru/svgogin/service/spark/web/SparkControllerItTest.java:
```diff
@@ -249,8 +249,169 @@ void saveCompanyShouldThrowExceptionWhenExists() throws Exception {
   }
+
+  @Test
+  void getCompanyByInnShouldThrowException_WhenPathInnIsNotValid() throws Exception {
```
why do you test it in integration test?

svgogin replied:
Tests have been moved to SparkControllerTest

svgogin committed:
Validation #23
7) Validation tests were moved to SparkControllerTest
8) Common multiple error response format implemented

alinkov commented on src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java:
```diff
@@ -1,38 +1,86 @@
 public class RestResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {
   private static final Logger log = LoggerFactory
       .getLogger(RestResponseEntityExceptionHandler.class);
+  ErrorCode errorCode;
```
why do you this?
This class is a singleton and can work in multiple threads. In this case you can return incorrect result

svgogin replied:
Fixed

alinkov commented on src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java:
```diff
@@ -198,20 +198,184 @@ void deleteCompanyShouldDeleteEntities() throws Exception {
+
+  @Test
+  void getCompanyByInnShouldThrowException_WhenPathInnIsNotValid() throws Exception {
+    // given
```
```suggestion
```
nit: You can remove this line

svgogin replied:
Done

alinkov commented on src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java:
```diff
@@ -198,20 +198,184 @@ void deleteCompanyShouldDeleteEntities() throws Exception {
+  }
+
+  @Test
+  void saveCompanyShouldThrowException_WhenOgrnInBodyIsNotValid() throws Exception {
```
```suggestion
  void saveCompanyShouldThrowException_WhenOgrnIsNotValid() throws Exception {
```

And also, you can use snake_case for tests:
```suggestion
  void save_company_should_throw_exception_when_ogrn_is_not_valid() throws Exception {
```

svgogin replied:
Ok

svgogin committed:
Validation #24
1) Issue with the singleton fixed.
2) Method toErrorDto overloaded.
3) /given removed where not used

alinkov commented on src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java:
```diff
@@ -20,19 +31,66 @@
-    String message = ex.getMessage();
+    var message = ex.getMessage();
+    var errorCode = ErrorCode.ERROR002;
+    var bodyOfResponse = List.of(new ErrorDto(errorCode, message));
```
nit: you can inline variables:
```suggestion
    var bodyOfResponse = List.of(new ErrorDto(ErrorCode.ERROR002, ex.getMessage()));
```
